File: test/cpp/harness.cpp
#include <iostream>
#include <sstream>
#include <memory>

#include <hoytech/error.h>
#include <hoytech/hex.h>

#include "negentropy.h"
#include "negentropy/storage/BTreeMem.h"
#include "negentropy/storage/Vector.h"



std::vector<std::string> split(const std::string &s, char delim) {
    std::vector<std::string> result;
    std::stringstream ss (s);
    std::string item;

    while (getline (ss, item, delim)) {
        result.push_back (item);
    }

    return result;
}



int main() {
    uint64_t frameSizeLimit = 0;
    if (::getenv("FRAMESIZELIMIT")) frameSizeLimit = std::stoull(::getenv("FRAMESIZELIMIT"));

    negentropy::storage::BTreeMem storage;
    std::unique_ptr<Negentropy> ne;

    std::string line;
    while (std::cin) {
        std::getline(std::cin, line);
        if (!line.size()) continue;

        auto items = split(line, ',');

        if (items[0] == "item") {
            if (items.size() != 3) throw hoytech::error("wrong num of fields");
            uint64_t created = std::stoull(items[1]);
            auto id = hoytech::from_hex(items[2]);
            storage.addItem(created, id);
        } else if (items[0] == "seal") {
            storage.seal();
            ne = std::make_unique<Negentropy>(storage, frameSizeLimit);
        } else if (items[0] == "initiate") {
            auto q = ne->initiate();
            if (frameSizeLimit && q.size() > frameSizeLimit) throw hoytech::error("initiate frameSizeLimit exceeded: ", q.size(), " > ", frameSizeLimit);
            std::cout << "msg," << hoytech::to_hex(q) << std::endl;
        } else if (items[0] == "msg") {
            std::string q;
            if (items.size() >= 2) q = hoytech::from_hex(items[1]);

            if (ne->isInitiator) {
                std::vector<std::string> have, need;
                auto resp = ne->reconcile(q, have, need);

                for (auto &id : have) std::cout << "have," << hoytech::to_hex(id) << "\n";
                for (auto &id : need) std::cout << "need," << hoytech::to_hex(id) << "\n";

                if (!resp) {
                    std::cout << "done" << std::endl;
                    continue;
                }

                q = *resp;
            } else {
                q = ne->reconcile(q);
            }

            if (frameSizeLimit && q.size() > frameSizeLimit) throw hoytech::error("frameSizeLimit exceeded: ", q.size(), " > ", frameSizeLimit, ": from ", (ne->isInitiator ? "initiator" : "non-initiator"));
            std::cout << "msg," << hoytech::to_hex(q) << std::endl;
        } else {
            throw hoytech::error("unknown cmd: ", items[0]);
        }
    }

    return 0;
}

File: test/go/harness.go
package main

import (
	"strings"
	"bufio"
	"fmt"
	"os"
	"strconv"
	"encoding/hex"
	negentropy "github.com/illuzen/go-negentropy"
)

func split(s string, delim rune) []string {
	return strings.FieldsFunc(s, func(r rune) bool {
		return r == delim
	})
}

func main() {
	frameSizeLimit := uint64(0)
	if env, exists := os.LookupEnv("FRAMESIZELIMIT"); exists {
		var err error
		frameSizeLimit, err = strconv.ParseUint(env, 10, 64)
		if err != nil {
			panic(fmt.Errorf("invalid FRAMESIZELIMIT: %w", err))
		}
	}

	storage := negentropy.NewVector()
	var ne *negentropy.Negentropy

	scanner := bufio.NewScanner(os.Stdin)
	const maxCapacity = 1024 * 1024 // 1MB
	buf := make([]byte, maxCapacity)
	scanner.Buffer(buf, maxCapacity)

	for scanner.Scan() {
		line := scanner.Text()
		if len(line) == 0 {
			continue
		}

		items := split(line, ',')

		switch items[0] {
		case "item":
			if len(items) != 3 {
				panic("wrong num of fields")
			}
			created, err := strconv.ParseUint(items[1], 10, 64)
			if err != nil {
				panic(err)
			}
			id, err := hex.DecodeString(items[2]) // Assume fromHex translates hex string to []byte
			if err != nil {
				panic(err)
			}
			storage.Insert(created, id)

		case "seal":
			storage.Seal()
			neg, err := negentropy.NewNegentropy(storage, frameSizeLimit)
			if err != nil {
				panic(err)
			}
			ne = neg

		case "initiate":
			q, err := ne.Initiate()
			if err != nil {
				panic(err)
			}
			if frameSizeLimit != 0 && uint64(len(q)) > frameSizeLimit {
				panic("initiate frameSizeLimit exceeded")
			}
			fmt.Printf("msg,%s\n", hex.EncodeToString(q))

		case "msg":
			var q []byte
			if len(items) >= 2 {
				s, err := hex.DecodeString(items[1])
				if err != nil {
					panic(err)
				}
				q = s
			}

			if (*ne).IsInitiator {
				var have, need []string
				resp, err := ne.ReconcileWithIDs(q, &have, &need)
				if err != nil {
					panic(fmt.Sprintf("Reconciliation failed: %v", err))
				}

				for _, id := range have {
					fmt.Printf("have,%s\n", hex.EncodeToString([]byte(id)))
				}
				for _, id := range need {
					fmt.Printf("need,%s\n", hex.EncodeToString([]byte(id)))
				}

				if resp == nil {
					fmt.Println("done")
					continue
				}

				q = resp
			} else {
				s, err := ne.Reconcile(q)
				if err != nil {
					panic(fmt.Sprintf("Reconciliation failed: %v", err))
				}
				q = s
			}

			if frameSizeLimit > 0 && uint64(len(q)) > frameSizeLimit {
				panic("frameSizeLimit exceeded")
			}
			fmt.Printf("msg,%s\n", hex.EncodeToString(q))

		default:
			panic("unknown cmd: " + items[0])
		}
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}
}